src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -5258,6 +5258,194 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
+                throw SemanticError("str.capitalize() takes no arguments",
+                    loc);
+            }
+            if (s_var.length() > 0) {
+                s_var[0] = toupper(s_var[0]);
+            }
+        } else if (attr_name == "lower") {
+            if (args.size() != 0) {
+                throw SemanticError("str.lower() takes no arguments",
+                        loc);
+            }
+            for (auto &i : s_var) {
+                if (i >= 'A' && i<= 'Z') {
+                    i = tolower(i);
+                }
+            }
+        } else if (attr_name == "find") {
+            if (args.size() != 1) {
+                throw SemanticError("str.find() takes one arguments",
+                        loc);
+            }
+            ASR::expr_t *arg = args[0].m_value;
+            ASR::ttype_t *type = ASRUtils::expr_type(arg);
+            if (!ASRUtils::is_character(*type)) {
+                throw SemanticError("str.find() takes one arguments of type: str",
+                    arg->base.loc);
+            }
+            if (ASRUtils::expr_value(arg) != nullptr) {
+                ASR::StringConstant_t* sub_str_con = ASR::down_cast<ASR::StringConstant_t>(arg);
+                std::string sub = sub_str_con->m_s;
+                //KMP matching
+                int str_len = s_var.size();
+                int sub_len = sub.size();
+                bool flag = 0;
+                int res = -1;
+                std::vector<int> lps(sub_len, 0);
+                if (str_len == 0 || sub_len == 0) {
+                    res = (!sub_len || (sub_len == str_len))? 0: -1;
+                } else {
+                    for(int i = 1, len = 0; i < sub_len;) {
+                        if (sub[i] == sub[len]) {
+                            lps[i++] = ++len;
+                        } else {
+                            if (len != 0) {
+                                len = lps[len - 1];
+                            } else {
+                                lps[i++] = 0;
+                            }
+                        }
+                    }
+                    for (int i = 0, j = 0; (str_len - i) >= (sub_len - j) && !flag;) {
+                        if (sub[j] == s_var[i]) {
+                            j++, i++;
+                        }
+                        if (j == sub_len) {
+                            res = i - j;
+                            flag = 1;
+                            j = lps[j - 1];
+                        } else if (i < str_len && sub[j] != s_var[i]) {
+                            if (j != 0) {
+                                j = lps[j - 1];
+                            } else {
+                                i = i + 1;
+                            }
+                        }
+                    }
+                }
+                tmp = ASR::make_IntegerConstant_t(al, loc, res,
+                    ASRUtils::TYPE(ASR::make_Integer_t(al,loc, 4, nullptr, 0)));
```

Comment:
Please put this into a function and call it here.

Reply:
Done